fdbserver/ClusterController.actor.cpp

```diff
@@ -223,6 +223,243 @@ bool ClusterControllerData::remoteTransactionSystemContainsDegradedServers() {
+		    worker.details.processClass.machineClassFitness(ProcessClass::LogRouter) < ProcessClass::NeverAssign) {
+			workers.push_back(worker.details);
+		}
+	}
+
+	if (workers.empty()) {
+		TraceEvent(SevWarn, "NoWorkersForLogRouter", cluster->id)
+		    .detail("TagId", logRouterTagId)
+		    .detail("TargetDcId", targetDcId);
+		throw recruitment_failed();
+	}
+
+	state WorkerInterface worker = workers[deterministicRandom()->randomInt(0, workers.size())].interf;
+
+	TraceEvent("RecruitingLogRouterOnWorker", cluster->id)
+	    .detail("WorkerID", worker.id())
+	    .detail("TagId", logRouterTagId)
+	    .detail("TargetDcId", targetDcId);
+
+	// Step 3: Initialize the log router
+	state InitializeLogRouterRequest req;
+	req.recoveryCount = db->recoveryData->cstate.myDBState.recoveryCount;
+	req.routerTag = routerTag;
+	req.startVersion = startVersion;
+	req.locality = config.tLogs[logSetIndex].locality;
+	req.isReplacement = true;
+
+	for (auto& tLogSet : config.tLogs) {
+		if (tLogSet.isLocal && tLogSet.tLogs.size() > 0) {
+			req.tLogLocalities = tLogSet.tLogLocalities;
+			req.tLogPolicy = tLogSet.tLogPolicy;
+			break;
+		}
+	}
+
+	if (db->recoveryData->logSystem.isValid()) {
+		auto lsConfig = db->recoveryData->logSystem->getLogSystemConfig();
+		if (lsConfig.recoveredAt.present()) {
+			req.recoverAt = lsConfig.recoveredAt.get();
+		}
+	}
+
+	req.allowDropInSim = false;
+
+	state TLogInterface logRouterInterf = wait(transformErrors(
+	    throwErrorOr(worker.logRouter.getReplyUnlessFailedFor(req,
+	                                                           SERVER_KNOBS->TLOG_TIMEOUT,
+	                                                           SERVER_KNOBS->MASTER_FAILURE_SLOPE_DURING_RECOVERY)),
+	    recruitment_failed()));
+
+	TraceEvent("LogRouterRecruited", cluster->id)
+	    .detail("TagId", logRouterTagId)
+	    .detail("LogRouterID", logRouterInterf.id())
+	    .detail("WorkerID", worker.id())
+	    .detail("LogSetIndex", logSetIndex);
+
+	// Step 4: Update the log system configuration
+	logSystem->updateLogRouter(logSetIndex, logRouterTagId, logRouterInterf);
+
+	// Step 5: Trigger registration update to propagate to ServerDBInfo
+	db->recoveryData->registrationTrigger.trigger();
+
+	TraceEvent("LogRouterRecruitmentComplete", cluster->id)
+	    .detail("TagId", logRouterTagId)
+	    .detail("LogRouterID", logRouterInterf.id());
+
+	return Void();
+}
+
+ACTOR Future<Void> monitorAndRecruitLogRouters(ClusterControllerData* cluster, ClusterControllerData::DBInfo* db) {
+	// Wait until log system is valid
+	while (!db->recoveryData.isValid() || !db->recoveryData->logSystem.isValid()) {
+		wait(delay(1.0));
+	}
+
+	state Reference<ILogSystem> logSystem = db->recoveryData->logSystem;
+	state uint64_t recoveryCount = db->recoveryData->cstate.myDBState.recoveryCount;
+	state LogSystemConfig config = logSystem->getLogSystemConfig();
+
+	// Find the log set with log routers (should be remote/satellite)
```

Comment:
Is it better to do the population of "logSetIndex" below where you are re-checking and setting "logSetIndex" in case config changed?